### PR TITLE
all: Send deadlocks as failures, crash only as a last resort

### PR DESCRIPTION
**Pull request by imsodin**
https://github.com/syncthing/syncthing/issues/7764

> I used to regularly check f/pmut "deadlocks" but all my samples look like false positives. Actually I think we should make our reaction less painful: What do you think about having the same kind of timeout as now, but it only results in a failure event (which I'd extend to be able to transport goroutine traces), and only after a much longer time do we actually panic. I'd still keep the panic for cases like e.g. introducing a real deadlock in a new RC release, which would otherwise require a manual upgrade (or at least restart) to resolve. My hope is to reduce the amount of false positive lock-panics and thus make it less likely for goleveldb to fail and even if the db is fine, it's also nicer on devices already being slow, if we don't interrupt their slow progress every now and then.

Now after the timeout, we send a failure report. Then we wait for 10 times the timeout, after which we send another failure report, try to shut down and only if that doesn't work (after all, it might indeed be a deadlock), panic.

To make those failure reports actually helpful, I added a goroutine field to failure reports. Those are saved on the crash-receiver server similarly to how crash reports are saved now. I also added a generic string map for extraneous info that might be added to a failure report (used for timeout length in deadlock detection).

### CODE REVIEW

**Comment by calmh** [comment]
I'm not convinced we need something this elaborate... Do we get a lot of deadlock failure reports? I can agree on removing the crash given most of the reports are false alarms, instead just keeping failure report plus a warning in the log and leaving it at that... If we did that we could enable it in release builds as well.

**Comment by imsodin** [comment]
There are not many deadlock reports any more, 0-10 per day in the last 14 days.

My motivation for extending the failure reports with goroutines is two-fold:  
I wanted goroutine traces for unclean shutdowns (services or database doesn't terminate within timeout) - there may be more instances. And now deadlock reports without goroutine traces would be pretty much useless. I did find a lot of locking improvements through deadlock reports in the past. Granted, there haven't been any useful reports in quite a while.

And independent of that I'd like to retain the crashing in candidate releases, in case we introduce a new deadlock: Without crashing the user needs to manually intervene to get unstuck, instead of Syncthing upgrading itself after the crash.

So in my opinion there's two options here:

1. The elaborate one with goroutines, i.e. PR status quo.
2. Keep crashing like we do now, just increase the timeout (because I really don't think we should crash after only 20min).

Opinions? Am I missing a "middle-ground" step here?

File: lib/model/util.go [hunk 1]
```diff
@@ -13,57 +13,95 @@ import (
...
-						msg += "\n===" + otherName + "===\n" + otherHolder.Holders()
-					}
+			select {
+			case <-warn.C:
+				failure := ur.FailureDataWithGoroutines(fmt.Sprintf("potential deadlock detected at %s (short timeout)", name))
+				failure.Extra["timeout"] = d.warnTimeout.String()
+				d.evLogger.Log(events.Failure, failure)
+			case <-done:
+				warn.Stop()
+				fatal.Stop()
```
Comment (calmh) [inline lib/model/util.go:66]: Maybe make the inner loop a method so we can defer these on creation?

File: lib/model/util.go [hunk 2]
```diff
@@ -13,57 +13,95 @@ import (
...
+				failure := ur.FailureDataWithGoroutines(err.Error())
+				failure.Extra["timeout"] = d.fatalTimeout.String()
+				var others string
+				for otherName, otherHolder := range d.otherHolders() {
+					others += "===" + otherName + "===\n" + otherHolder + "\n"
 				}
-				panic(msg)
+				others = others[:len(others)-1]
+				failure.Extra["other-holders"] = others
+				d.otherHolders()
```
Comment (calmh) [inline lib/model/util.go:81]: This is called again for some side effect...?

File: lib/model/util.go [hunk 3]
```diff
@@ -13,57 +13,95 @@ import (
...
+			case <-fatal.C:
+				err := fmt.Errorf("potential deadlock detected at %s (long timeout)", name)
+				failure := ur.FailureDataWithGoroutines(err.Error())
+				failure.Extra["timeout"] = d.fatalTimeout.String()
+				var others string
+				for otherName, otherHolder := range d.otherHolders() {
+					others += "===" + otherName + "===\n" + otherHolder + "\n"
 				}
-				panic(msg)
+				others = others[:len(others)-1]
```
Comment (calmh) [inline lib/model/util.go:79]: What are we cutting off the end here?

File: lib/model/util.go [hunk 4]
```diff
@@ -13,57 +13,95 @@ import (
...
+				time.Sleep(time.Minute)
+				panic(fmt.Sprintf("%v:\n%v", err, others))
+			case <-done:
+				fatal.Stop()
 			}
 		}
 	}()
 }
 
+func (d *deadlockDetector) otherHolders() map[string]string {
```
Comment (calmh) [inline lib/model/util.go:95]: Looks like this might as well return a single formatted string if that's how we always treat the resulting map

**Review by calmh — COMMENTED** [review]
I buy the argument for the elaborateness

File: lib/model/util.go (hunk 3 above)
Comment (imsodin) [reply lib/model/util.go:79]: An empty line.

**Review by calmh — APPROVED** [review]

**Pull request merged**